Custom extensions

2bndy5 opened the pull request:
# New Directives
These new directives are automatically added to doc builds because they are tied to our CSS:

- `md-tab-set` and `md-tab-item` which resolves #43 (includes a new doc page to demonstrate CSS `class`/`id` usage)
- `md-mermaid` to expose the implementation merged in from upstream (includes a new doc page to demonstrate CSS `class`/`id` usage).

# Patches
- monkey-patch for the `sphinxcontrib-details-directive` which is applied only when the directive is installed. This patch allows us to use the `:class:` option to specify styling for collapsible admonitions. resolves #15 again
- updates _package.lock_ from changes in #44 
- auto-asserts line numbers as a table's column in code-blocks - resolves #26 until sphinx v6.0 is released
- removes dependency on `sphinx-design` in docs' builds, but adds `sphinxcontrib-details-directive` to the docs' extensions/reqs.

## Known bug (very minor annoyance)
Upon page refresh, the tabs' highlighted bottom border always resets to the first tab (unexpected behavior), but the displayed tab content and highlighted tab label do not reset (expected behavior). I suspect it has something to do with how the JS loads on page refresh, but I'm not too concerned at this point.

### before refresh
![image](https://user-images.githubusercontent.com/14963867/158981622-e48b5357-f7c5-4999-ba8c-c770756afb1d.png)

### after refresh
![image](https://user-images.githubusercontent.com/14963867/158981768-13ceec9f-7c93-4261-ba33-7246955d688f.png)

2bndy5 commented:
I've started seeing a deprecation warning when executing `pip setup.py bdist_wheel` cmd:
```
/path/to/sphinx-immaterial/setup.py:21: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
  import distutils.command.build
```
Seems to only appear in python v3.10+ or could be caused by pip v22.0.4

jbms commented on docs/content_tabs.rst:
```diff
@@ -0,0 +1,122 @@
+============
+
+.. note::
+    This document focused on content tabs, not navigation tabs.
```
"This document discusses content tabs, not navigation tabs."

jbms commented on docs/content_tabs.rst:
```diff
@@ -0,0 +1,122 @@
+.. note::
+    This document focused on content tabs, not navigation tabs.
+
+Use of `content tabs in mkdocs-material <https://squidfunk.github.io/mkdocs-material/reference/content-tabs/>`_
```
"in the mkdocs-material"

jbms commented on docs/content_tabs.rst:
```diff
@@ -0,0 +1,122 @@
+.. admonition:: Linked Tabs
+    :class: missing
+
+    The `linked content tabs <https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#linked-content-tabs>`_
```
"feature seen in mkdocs-material"

"to its public releases."

jbms commented on docs/content_tabs.rst:
```diff
@@ -0,0 +1,122 @@
+    seen in mkdocs-material is not supported until that feature transitions from the mkdocs-material theme's insider
+    releases to it's public releases.
+
+    However, you can use other sphinx extensions (like `sphinx-design tabs`_) to achieve this functionality.
```
"You can use other..."  "However, other extensions "

jbms commented on docs/content_tabs.rst:
```diff
@@ -0,0 +1,122 @@
+    Although, other extensions will require some custom CSS styling to match the mkdocs-material
+    theme's styling for content tabs.
+
+.. confval:: md-tab-set
```
Rather than confval, we should use the rst domain provided by sphinx:

https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#the-restructuredtext-domain

2bndy5 replied:
Done.

This is the first time I've actually used the rst domain, so let me know if there are any tweaks you'd like to see,

jbms commented on docs/mermaid_diagrams.rst:
```diff
@@ -0,0 +1,207 @@
+.. note::
+    Use of this feature has no affect or affiliation with sphinx's graphviz implementation.
+
+The mkdocs-material theme is equipped to make use of diagrams generated (during page load time)
```
Perhaps put this as a `.. note` at the bottom, since mkdocs-material's implementation isn't terribly relevant to users of this theme, except I suppose if they are migrating.

2bndy5 replied:
Its more of an introductory note. I could hide it into a collapsible admonition titled "Implementation Note".

There is a similar introductory statement in practically all supported MD extensions (in mkdocs) that get special attention for this port of the theme. This was added specifically for those migrating from mkdocs.

jbms replied:
I'll leave this up to your judgement then.

2bndy5 replied:
I favor leaving it as is because a collapsible admonition will still draw attention (even if collapsed by default). If the implementation detail was much longer or more specific, then it would be a good idea to hide that in a collapsible admonition.

Notice that this paragraph exists before the directive's docs, so a link directly to the directive's docs will scroll past the introductory note(s).

2bndy5 replied:
That and I don't want to retrofit all the other statements that talk about the differences between upstream and this port.

jbms commented on docs/mermaid_diagrams.rst:
```diff
@@ -0,0 +1,207 @@
+
+.. code-block:: rst
+
+    .. md-mermaid::
```
Note: it might be nice to add a directive for us in our documentation that avoids having to duplicate the code for samples like this.

2bndy5 replied:
The `sphinx-design` docs use external files to store the duplicated snippets. The external files are also used in test runs via `pytest`. Their docs are a bit more laborious because they also support the myRST parser (which involves duplicated markdown versions of rst snippets).

jbms replied:
Yeah it was just a suggestion --- I didn't imagine it as an extension provided by this theme to users, but perhaps just something defined for our own documentation, like `confval`.  But it is true it is widely useful for other sphinx extensions, though maybe some customization is needed for different uses.

2bndy5 replied:
After giving this more thought, IDK if an external extension is practical for everyone's needs (multiple example snippets can rendered identically). I'll try to keep the idea on my radar, but I doubt I'll ever implement something like this.

jbms commented on setup.py:
```diff
@@ -17,20 +17,19 @@
 # IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 # WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
-import setuptools
```
setuptools should always be imported first because it does some monkey patching of distutils.

We should add a comment to indicate that.

2bndy5 replied:
distutils is going away; see https://github.com/jbms/sphinx-immaterial/pull/45#issuecomment-1072288302

until we phase out the use of distutils, I could add a pylint comment to ignore the import ordering, here.

jbms replied:
I think distutils isn't exactly going away, it is just being moved out of the Python standard library into setuptools itself.

Furthermore, I think setuptools modifies `sys.modules` so that you can still import distutils using the normal path:
https://github.com/pypa/setuptools/blob/main/_distutils_hack/__init__.py

Does the deprecation warning still appear if you fix the import order?

2bndy5 replied:
pylint realizes the setuptools is not in std lib, so it complains about the order. Using
```python

import setuptools  # pylint: disable=wrong-import-order

import atexit
import distutils.command.build
import os
import subprocess
import tempfile

import setuptools.command.build_py
import setuptools.command.develop
import setuptools.command.install
import setuptools.command.sdist
```
doesn't seem to trigger the deprecatin warning. 🤷🏼

2bndy5 replied:
After reading [backward compatibility in PEP 632](https://peps.python.org/pep-0632/#backwards-compatibility), I found that `setuptools` (as of v60.0) has import paths that help mitigate the removal of distutils. According to [the setuptools docs](https://setuptools.pypa.io/en/latest/deprecated/distutils-legacy.html#prefer-setuptools), we should use 
```ini
# pyproject.toml
[build-system]
requires = ["setuptools>=60", "wheel", "setuptools_scm"]
```
```python
# setup.py
try:
    import setuptools.command.build
except ImportError:  # for setuptools<60.0
    import distutils.command.build
```

jbms replied:
I don't think `setuptools.command.build` is present in setuptools 60.  It isn't listed on that webpage and I tried installing setuptools 60.10.0 and get an import error.

However, as long as we import setuptools first we are in fact using the version of distutils from setuptools, which is why there is no deprecation warning.

2bndy5 replied:
Strange, that deprecation warning is completely useless then. All it did was cause confusion here.

jbms commented on sphinx_immaterial/mermaid_diagrams.py:
```diff
@@ -0,0 +1,59 @@
+
+
+def visit_mermaid_node_latex(self, node: mermaid_node):
+    self.body.append('<pre class="mermaid">')
```
This doesn't work, right?

This theme isn't intended to be used except with the html builder, so I don't think we need to worry about latex.

2bndy5 replied:
It does work with latex writer. That's why I have the CI building both html and latex. In fact the RTD builds are making PDFs from the latex builds.

The mermaid feature won't work with latex since it relies on JS. In fact if the JS in a browser isn't loaded, then the mermaid directives look like non-highlighted code blocks. This is what I'm aiming for with the latex writer.

This function is important because the `starttag()` used in html doesn't exist in the latex builder.

jbms replied:
Oh I see ---  putting the mermaid code literally in the output seems reasonable enough for latex (rendering the diagrams would be better but of course that would be a lot more trouble to accomplish), but why include HTML like `<pre class="mermaid">`?

2bndy5 replied:
that's just what it does for html. I don't really know latex syntax...

Actually the mermaid directive results in a
```html
<div class="mermaid-diagram">
  <pre class="mermaid"> <!--this part is reuired for the mermaid.js to work-->
    <code>
      <!-- all special markup language characters are converted -->
    </code>
  </pre>
</div>
```

2bndy5 replied:
I installed all the `tex-*` stuff used in the RTD docker and test built a pdf from the latex output... The HTML syntax used here (for latex writers) is now replaced with latex syntax to better imitate a literal code block in the generated PDF.

As a side note, I don't think that PDFs are built on RTD because of other numerous unknown references (seems related to `.. _ref_id:` tactic). A PDF can be built from this theme's latex output, but you have to accept that there's no ToC written. We can address this in the future if someone is so inclined, but it isn't the primary goal of this theme. So, for now, I'm just doing the bare minimum to allow a latex builder to complete its job.

jbms commented on sphinx_immaterial/content_tabs.py:
```diff
@@ -0,0 +1,220 @@
+def setup(app: Sphinx) -> None:
+    app.add_directive("md-tab-set", MaterialTabSetDirective)
+    app.add_directive("md-tab-item", MaterialTabItemDirective)
+    app.add_post_transform(MaterialTabSetHtmlTransform)
```
I think it would be best to avoid using `add_post_transform`, since that means an additional full traversal of every document.  I think the main thing you are using the post transform for is assigning a unique index to each tab set --- for that you could instead use `env.temp_data` (which is cleared for each document) or handle it in the writer (i.e. the `visit_` functions), where you could store the counter as an attribute on the writer.

2bndy5 replied:
>  I think the main thing you are using the post transform for is assigning a unique index to each tab set

Not true. The transformer, takes the tab item and splits it up into a label, input, and content's div elements. For each tab set the labels and and content's div elements go in their own div containers, while the input elements go directly after the div for the tab set but before the label and content groups.

I couldn't think of a way to organize these elements from the `MaterialTabItemDirective.run()` And, there was the additional pain of figuring out the `id`, `for`, and `name` attributes value; I didn't know about the `env.temp_data`.

jbms replied:
Okay, I see.  Thinking about this more, I think this could all be handled in `visit_tab_set`:

- Store the counter for ids in the HTMLTranslator (i.e. `self`)
- In `visit_tab_set`, emit the opening div, then emit the input tag for each child, then emit the opening tag for the labels.  Then emit the opening label tag for each child, and call `child_label.walkabout(self)` to render the label.
- After emitting the labels, do a similar thing for the content.
- Emit the closing tag.
- raise docutils.nodes.SkipNode to prevent docutils from itself recursing on the children.

2bndy5 replied:
Interesting challenge. Much of this code is heavily modified from the `sphinx-design` ext. I didn't bother with attribution because the `sphinx-design` uses MIT lisence, and the code changed so much from the initial copy.

Clearly, I have a lot to learn about `docutils` - I wish they had better/complete docs.

jbms replied:
Yeah unfortunately the docs are rather limited --- I have found the only way to really understand docutils and sphinx is to study the source code.

2bndy5 replied:
Agreed, its just **_a lot_** to study. Also, its not a good sign when a docs-based library lacks on its own docs... 🤣

2bndy5 replied:
`temp_data` persists across all pages, so storing the tab set count in `temp_data` yields an `id` of `__tabbed_4` on the second set of tabs in content_tabs.rst. This is not consistent with upstream HTML generation. It doesn't seem to be a problem, but it is a less precise re-implementation. My main concern here is having to retool it when linked tabs moves to public releases in upstream.

### un-committed mod
Currently, I'm not using the `walkabout()` idea. I basically moved all the algorithm from the transformer into `visit_tab_set()`. Is there an objection to having separate `visit_*()` functions? I'm uneasy about writing all the raw HTML in a single `visit_*()` function.

jbms replied:
temp_data should be cleared after reading each document, but note that it only applies to the initial reading stage, i.e. when the directive is first being processed.  During later stages I don't think it gets reset, so it shouldn't be used for per-document state.  During writing, you can instead use the HTMLTranslator object itself (i.e. the `self` parameter of the visit methods) to store per-document state, because a separate HTMLTranslator object is created per document.

2bndy5 replied:
I can't treat the translator like a dict; that results in a error. I'm currently doing
```py

def visit_tab_set(self: HTMLTranslator, node: content_tab_set):
    # increment tab set counter
    temp_data = self.builder.env.temp_data
    temp_data["tab_set_count"] = temp_data.get("tab_set_count", 0) + 1
```
Even dynamic attribute creation has the same effect
```python
def visit_tab_set(self: HTMLTranslator, node: content_tab_set):
    # increment tab set counter
    self.tab_set_count = getattr(self, "tab_set_count", 0) + 1
```
The old way i was using was a global 1-item list in content_tabs.py that did get reset for each page.

2bndy5 replied:
NM, dynamic attribute creation worked I was just misreading the HTML... commit incoming.

jbms commented on sphinx_immaterial/content_tabs.py:
```diff
@@ -0,0 +1,220 @@
+        return [tab_item]
+
+
+class content_tab_input(nodes.Element, nodes.General):
```
I think it would make more sense to have a `content_tab_item` node, and then handle the input and label stuff in the html translation only.

2bndy5 replied:
I need the writer to explicitly use different tag names for the input & label elements. Are you saying there's a way to do that without the `stattag()` in a visitor function?

Also the attributes of a node aren't directly output as attributes of an element. These visitor functions take care of adding peculiar attributes like `for`,

jbms replied:
starttag just formats an HTML tag:
https://sourceforge.net/p/docutils/code/HEAD/tree/trunk/docutils/docutils/writers/_html_base.py#l411

But why can't you use it?

2bndy5 replied:
> But why can't you use it?

I think I worded my previous comment wrong. I need to use the `starttag()`. Is there a way to change the html tag without using `starttag()`?

jbms replied:
Basically the way the document gets converted to HTML is that the HTMLTranslator class (defined in docutils but extended by sphinx) has a `self.body` member which is just a list of strings that will get concatenated together to form the final HTML output..

Then docutils traverses the document using the walkabout function, using the HTMLTranslator object as a visitor.

In these visit methods you can just append any arbitrary text that you like, there isn't any restriction.  Additionally, in the `visit_xxx` function if you have already handled the entire content of the element you can raise `docutils.nodes.SkipNode` to prevent docutils from recursing on the children and calling the `depart_xxx` function.  So that allows you to have full control over the output.

The `starttag` method is just a utility function for formatting an HTML start tag so you can use it as you like, but you can also just format the HTML directly.  There is also a `self.attval` function for escaping an HTML attribute.

2bndy5 replied:
There seems to be a lot of convenient translation (like `<` into `&lt;`) and string normalization (for class names and id attributes) that I don't want to circumvent.

jbms replied:
If you want to use starttag and don't have a suitable node object to pass it, you could always just create a docutils node object on the fly.

2bndy5 replied:
Its hard to track what the request/comment here is concerning. Recent developments have improved the `visit_tab_set()` to factor out the `content_tab_input` class entirely. I think this should be resolved.

jbms commented:
Overall, great work here!

I think we should try to fix the indicator line beneath the tab.  Probably adding some console.log statements to the javascript code may help in figuring out what is going on.

jbms commented:
I looked into the indicator line issue -- turns out it is a bug in mkdocs-material which appears to have been fixed in the insiders version.

I reported it here:
https://github.com/squidfunk/mkdocs-material/issues/3744

jbms commented on sphinx_immaterial/content_tabs.py:
```diff
@@ -0,0 +1,207 @@
+        tab_item_identity = tab_set_identity + f"_{tab_total}"
+
+        # create: <input checked="checked" id="id" type="radio">
+        input_node = content_tab_input(
```
You could do this tag more simply as:

```
self.body.append(f'<input id="{self.attval(tab_item_identity)}" name="{self.attval(tab_set_identity)}" type="radio"{" checked" if tab_total == 1 else ""}>')
```

I think that more clearly indicates what is happening and avoids the need for the extra content_tab_input node.

2bndy5 replied:
Again, this assumes that what the user entered is well formed. I think by doing it the current way, it is clearer, and string normalization (including exceptions raised about invalid characters) from docutils is used.

I understand that there are other methods that can be called to take advantage of docutils conveniences, but this just seems like the simplest and conventional.

2bndy5 replied:
After giving this more thought, The conveniences of docutils in this instance shouldn't be needed. I applied this suggestion, but I still feel it isn't the most robust.

jbms commented on sphinx_immaterial/content_tabs.py:
```diff
@@ -0,0 +1,207 @@
+
+    # configure tab set's div attributes
+    tab_set_identity = "__tabbed_" + str(self.tab_set_count)
+    attributes = {"data-tabs": str(self.tab_set_count) + node["data-tabs"]}
```
Potentially rather than have an extra data-tabs attribute you could just compute that here from `len(node.children)`

2bndy5 replied:
I've found that there are multiple ways to count the children. I have no objection to this suggestion. I just set the attribute since I had to traverse the child nodes (in `Material TabSetDirective.run()`) to verify they were of `md-tab-item` directives (so I tallied the number of children there).

jbms commented on sphinx_immaterial/content_tabs.py:
```diff
@@ -0,0 +1,207 @@
+    # configure tab set's div attributes
+    tab_set_identity = "__tabbed_" + str(self.tab_set_count)
+    attributes = {"data-tabs": str(self.tab_set_count) + node["data-tabs"]}
+    node["set_id"] = tab_set_identity
```
What is `set_id` used for?

Normally the HTMLTranslator would not modify the document itself, though I am not sure if that is a problem --- e.g. I'm not clear if multiple translators would ever be run on the same copy of the document.

But in any case I think it would be simpler not to modify the node itself here or below.

Instead you can accomplish the same thing by calling e.g.:

`label_node.walkabout(self)`

That will cause all of the visit/depart methods to be called for the label_node.

2bndy5 replied:
`set_id` turns into the element's id attribute. IIRC, in docutils source, it gets concatenated with `ids` attribute. Though I think `set_id` takes precedence when given more the an 1 id (other ids - if they exist- get out into another attribute for the HTML element).

This function is setting up the necessary structure by re-writing the children nodes of a tab-set. There isn't an my need to call `walkabout()` (which would require another for loop at the end followed by a raised `SkipNode` exception). I feel like that just takes control away from docutils.

2bndy5 replied:
If I call `label_node.walkabout(self)` here then how do I use `SkipNode` on the child of a child from here?

2bndy5 replied:
ok I'm now calling
```py
    tab_label_div.walkabout(self)
    tab_content_div.walkabout(self)
    raise nodes.SkipNode()
```
instead of appending new children to the tab-set node.

This should completely satisfy this request.

jbms commented on sphinx_immaterial/content_tabs.py:
```diff
@@ -0,0 +1,182 @@
+        # create: <input checked="checked" id="id" type="radio">
+        self.body.append(
+            "<input "
+            + ('checked="checked"' if not tab_total else "")
```
It is also valid to just specify 'checked' --- for HTML boolean attributes it is not necessary to specify 'checked="checked"', though that is allowed:
https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes

In any case, though, I think you need a space afterwards, so that you don't end up with `<input checked="checked"type="radio" ...>`

2bndy5 replied:
I can add the space which doesn't seem to be needed as the generated docs perform just the same (probably because the ending quotation mark).

As for the `check="checked"`, I'm just trying to mimic what the HTML looks like on upstream. If you find that the JS doesn't care about this attribute's value, then I can change it to a boolean value.

Notice also that the attribute never changes when selecting different tabs.

2bndy5 replied:
Testing reveals that the quoted value is more resistant to missing spaces, but the boolean value works with the upstream JS just as well. I wish I could see the insider code so I won't have to retool this when that linked tabs feature goes public (~$900 away from that goal).

2bndy5 replied:
now using boolean value

jbms commented on sphinx_immaterial/content_tabs.py:
```diff
@@ -0,0 +1,182 @@
+    tab_total = 0
+    tab_label_div = nodes.container("", is_div=True, classes=["tabbed-labels"])
+    tab_content_div = nodes.container("", is_div=True, classes=["tabbed-content"])
+    for tab_item in node.children:
```
Might be slightly clearer to do:

`for i, tab_item in enumerate(node.children)`

rather than manually increment `tab_total`.

2bndy5 replied:
oh I forgot to change that.

jbms commented on sphinx_immaterial/content_tabs.py:
```diff
@@ -0,0 +1,182 @@
+        self.body.append(
+            "<input "
+            + ('checked="checked"' if not tab_total else "")
+            + f'type="radio" id="{tab_item_identity}" name="{tab_set_identity}">'
```
Note: You mentioned you were concerned about the robustness --- it is true that properly escaping attributes is an important issue when generating HTML.  Here both of these attributes are generated by this extension themselves and are known not to need escaping.  However, to make it extra robust you could use `self.attval(tab_item_identity)` and `self.attval(tab_set_identity)` --- `attval` is a method implemented by docutils that is used by the `starttag` method to properly escape attribute values.

2bndy5 replied:
thanks for the explanation! Adding it now.

jbms commented:
Thanks for making the changes --- it is true there are many ways to implement things and it is not always clear what the best approach is --- I think it is true that your initial approach of using a post transform was perfectly valid.  I just was recently trying to improve the efficiency of documentation generation for tensorstore, where due to the large number of pages generated for the api documentation, it can be rather slow, so I was acutely aware of efficiency concerns.  My thinking is also that ideally the docutils nodes represent semantically meaningful things, and then the HTML translator converts that to HTML.  In this case, the input elements aren't really a semantically meaningful thing on their own, they are just how it happens to be implemented in HTML, so it seems nice to encapsulate that in the HTML translator if possible.

jbms commented:
Regarding the indicator line issue -- if that doesn't get fixed in upstream mkdocs-material soon that should be easy to fix here.

I also think it would be fairly easy to implement the linked tabs feature if we want it, though I wonder if that should be enabled via a `:linked:` option or similar, rather than a global "feature" option.

2bndy5 commented:
> I also think it would be fairly easy to implement the linked tabs feature if we want it, though I wonder if that should be enabled via a :linked: option or similar, rather than a global "feature" option.

Whatever upstream does to do it would be preferable. If it ends up being harder to implement here, then I would also favor a `:linked:` option as that also _slightly_ resembles how sphinx-design does it.
- Rather than use the label, the sphinx-design uses a `:key: linked_label_x` that it assign to each `tab-item`. This is of course more long-winded, but using RST is more flexible than MD in adding a specific flag to a block.

jbms commented:
It wouldn't be hard to support it exactly as in mkdocs-material --- logically though it seems like it may be more useful to allow the user more flexibility.  As you note --- it is easier to support additional options in rst than in markdown.

jbms commented:
Thanks!

2bndy5 commented:
Either way, we'd have to brush up on rxjs syntax if the `:linked:` approach is added in with upstream's global implementation (in the same TS file). If I'm being honest, I feel like you're better equipped to make that kind of change. 😉

2bndy5 commented:
I still don't fully understand pushing tags directly. I tagged the merge commit https://github.com/jbms/sphinx-immaterial/commit/f925a444a9ca724d46cbacdf909e376ee2c68364 with v0.4.0, but the CI didn't seem to run on that commit.

jbms commented:
Yeah I can take a look at adding the linked support.

I am bit confused by what is going on with github actions --- we have seen issues before with it not triggering on this repository.  The config is identical to what I have on other repositories where it seems to work reliably, so I'm not sure what is going on here.

jbms commented:
Actually I'm not seeing your tag.  After you create a tag locally, you have to specifically push the tags to the remote repository:

```
git push --tags origin
```

2bndy5 commented:
Its got to related to PRs coming from other forks. I just switched my remotes to only use this repo from here on out. They recently added [a new event (for security concerns) specific to this situation](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target).

You can also use `gh` CLI to [manually trigger a workflow](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch).

RTD updated the "stable" build on the merging of this PR, so I think its related to event.

>Actually I'm not seeing your tag. After you create a tag locally, you have to specifically push the tags to the remote repository:

I did, but then I removed it when I realized the CI wasn't triggered for main.

---

I'm going to try adding a `workflow_dispatch` to the CI's list of triggered events. then tag it with v0.4.0 again.

2bndy5 commented:
well I got it to publish, but it keeps triggering twice
1. for commit
2. for the tag

This might be problematic for some scenarios of bumped version numbers. As witnessed when I pushed v0.3.1 tag:
```
Checking dist/sphinx_immaterial-0.3.post1.dev2-py3-none-any.whl: PASSED, with warnings
  warning: `long_description_content_type` missing. defaulting to `text/x-rst`.
Uploading distributions to https://test.pypi.org/legacy/
Uploading sphinx_immaterial-0.3.post1.dev2-py3-none-any.whl

  0%|          | 0.00/4.54M [00:00<?, ?B/s]
 85%|████████▌ | 3.86M/4.54M [00:00<00:00, 14.8MB/s]
100%|██████████| 4.54M/4.54M [00:01<00:00, 4.19MB/s]
HTTPError: 400 Bad Request from https://test.pypi.org/legacy/
Error during upload. Retry with the --verbose option for more details.
File already exists. See https://test.pypi.org/help/#file-name-reuse for more information.
```

jbms commented:
Yeah, it will trigger twice, but I think that isn't too big of a deal.  The error you are seeing is for an upload to the *test* PyPI instance.  Only the action run triggered by the tag push will upload to the non-test PyPI instance.

2bndy5 commented:
Unfortunately, such an error will cause the badge on the readme to indicate the failure to upload to test-pypi. I believe we can tell the workflow to [ignore the error for that specific step](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error)
